ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/altair/statetransition/epoch/ValidatorStatusFactoryAltair.java

```diff
@@ -52,33 +53,86 @@ protected void processParticipation(
     final BeaconStateAltair state = BeaconStateAltair.required(genericState);
 
     final SszList<SszByte> previousParticipation = state.getPreviousEpochParticipation();
-    for (int i = 0; i < previousParticipation.size(); i++) {
-      final SszByte participationFlags = previousParticipation.get(i);
-      if (miscHelpersAltair.hasFlag(
-          participationFlags.get(), ParticipationFlags.TIMELY_SOURCE_FLAG_INDEX)) {
-        statuses.get(i).updatePreviousEpochSourceAttester(true);
-      }
-      if (miscHelpersAltair.hasFlag(
-          participationFlags.get(), ParticipationFlags.TIMELY_TARGET_FLAG_INDEX)) {
-        statuses.get(i).updatePreviousEpochTargetAttester(true);
+    final SszList<SszByte> currentParticipation = state.getCurrentEpochParticipation();
+    for (int i = 0; i < statuses.size(); i++) {
+      final ValidatorStatus status = statuses.get(i);
+
+      if (status.isActiveInPreviousEpoch()) {
+        final byte previousParticipationFlags = previousParticipation.get(i).get();
+        if (miscHelpersAltair.hasFlag(
+            previousParticipationFlags, ParticipationFlags.TIMELY_SOURCE_FLAG_INDEX)) {
+          status.updatePreviousEpochSourceAttester(true);
+        }
+        if (miscHelpersAltair.hasFlag(
+            previousParticipationFlags, ParticipationFlags.TIMELY_TARGET_FLAG_INDEX)) {
+          status.updatePreviousEpochTargetAttester(true);
+        }
+        if (miscHelpersAltair.hasFlag(
+            previousParticipationFlags, ParticipationFlags.TIMELY_HEAD_FLAG_INDEX)) {
+          status.updatePreviousEpochHeadAttester(true);
+        }
       }
-      if (miscHelpersAltair.hasFlag(
-          participationFlags.get(), ParticipationFlags.TIMELY_HEAD_FLAG_INDEX)) {
-        statuses.get(i).updatePreviousEpochHeadAttester(true);
+
+      if (status.isActiveInCurrentEpoch()) {
+        final byte currentParticipationFlags = currentParticipation.get(i).get();
+        if (miscHelpersAltair.hasFlag(
+            currentParticipationFlags, ParticipationFlags.TIMELY_SOURCE_FLAG_INDEX)) {
+          status.updateCurrentEpochSourceAttester(true);
+        }
+        if (miscHelpersAltair.hasFlag(
+            currentParticipationFlags, ParticipationFlags.TIMELY_TARGET_FLAG_INDEX)) {
+          status.updateCurrentEpochTargetAttester(true);
+        }
       }
     }
+  }
```

Comment:
Behaviour here didn't actually change but we now only iterate through the validator set once instead of twice. We can also skip accessing the participation flags entirely for any validators that weren't active in that epoch as they can't have attested (and matches the spec better which iterates through the active validators for the epoch).